dsn.go

```diff
@@ -0,0 +1,31 @@
+package gorm
+
+import (
+	"net/url"
+	"strconv"
+)
+
+type DSN struct {
+	Host    string
+	Port    int
+	User    string
+	Pass    string
+	Db      string
+	Options map[string]string
+}
+
+func (d DSN) String() string {
+	dsn := d.User + ":" + d.Pass + "@tcp(" + d.Host + ":" + strconv.Itoa(d.Port) + ")/" + d.Db
```

Comment:
need to check whether the parameter is empty?

Reply:
I have tried, and if dsn is not valid, it will panic at mysql.Open.

There are no syntax problems because of the default values.

If you want to add a empty check, then you have to return error, so it doesn't iterate.

**Therefore, I recommend not adding empty check here.**